src/generate/rust.rs

```diff
@@ -1481,12 +804,549 @@ impl<Pat: Ord + Hash + RustInputPat> Rule<Pat> {
                         i,
                         rule.parse_node_kind(parse_nodes),
                         rule.generate_traverse_shape(true, parse_nodes)
-                    );
+                    ).unwrap();
                 }
-                write!(s, " }}");
+                write!(s, " }}").unwrap();
                 s
             }
             Rule::Opt(rule) => format!("[{}]", rule.generate_traverse_shape(true, parse_nodes)),
         }
     }
 }
+
+fn impl_parse_with(name: &str, rust_slice_ty: &str, out: &mut String) {
+    put!(out, "
+    impl<'a, 'i, I: ::gll::runtime::Input<Slice = ", rust_slice_ty, ">> ", name, "<'a, 'i, I> {
+        pub fn parse_with<R>(
+            input: I,
+            f: impl for<'b, 'i2> FnOnce(
+                &'b ::gll::runtime::Parser<'i2, _P, _C, I>,
+                ParseResult<'b, 'i2, I, ", name, "<'b, 'i2, I>>,
+            ) -> R,
+        ) -> R {
+            ::gll::runtime::Parser::with(input, |mut parser, range| {
+                let call = Call {
+                    callee: ", CodeLabel(name.to_string()), ",
+                    range,
+                };
+                parser.threads.spawn(
+                    Continuation {
+                        code: call.callee,
+                        fn_input: call.range,
+                        state: 0,
+                    },
+                    call.range,
+                );
+                parse(&mut parser);
+                let result = parser.memoizer.longest_result(call);
+                f(&parser, result.ok_or(ParseError::NoParse).and_then(|range| {
+                    let handle = Handle {
+                        node: ParseNode { kind: ", ParseNodeKind(name.to_string()), ", range },
+                        parser: &parser,
+                        _marker: PhantomData,
+                    };
+                    if range == call.range {
+                        Ok(handle)
+                    } else {
+                        Err(ParseError::TooShort(handle))
+                    }
+                }))
+            })
+        }
+    }")
+}
+
+fn declare_rule<Pat>(
```

Comment:
This function is yuuge... (^,-) would be nice to split it up into more units... :)

Reply:
Done.